signer/src/ecdsa.rs

```diff
@@ -8,85 +8,128 @@
+    pub signature: RecoverableSignature,
 }
 
-impl<T> Signed<T>
-where
-    T: ProtoSerializable + Clone,
-    T: Into<<T as ProtoSerializable>::Message>,
-{
-    /// Verify the signature over the inner data.
-    pub fn verify(&self) -> bool {
-        let msg = secp256k1::Message::from_digest(self.inner.digest());
-        let Ok(sig) = secp256k1::ecdsa::Signature::from_compact(&self.signature) else {
-            return false;
-        };
+impl Signed<SignerMessage> {
+    /// Determines the public key for which `sig` is a valid signature for
+    /// `msg`.
+    pub fn recover_ecdsa(&self) -> Result<PublicKey, Error> {
+        let msg = secp256k1::Message::from_digest(self.inner.to_digest());
+        self.signature.recover_ecdsa(&msg)
+    }
 
-        self.signer_pub_key.verify(SECP256K1, &msg, &sig).is_ok()
+    /// Determines the public key for which `sig` is a valid signature for
+    /// `msg`.
+    pub fn recover_ecdsa_with_digest(&self, digest: [u8; 32]) -> Result<PublicKey, Error> {
+        let msg = secp256k1::Message::from_digest(digest);
+        self.signature.recover_ecdsa(&msg)
     }
 
     /// Unique identifier for the signed message
     pub fn id(&self) -> [u8; 32] {
-        self.inner.digest()
+        self.inner.to_digest()
+    }
+
+    /// Decodes an encoded protobuf message, transforming it into a
+    /// [`Signed<SignerMessage>`], returning the signed message with the
+    /// digest that was signed.
+    ///
+    /// # Notes
+    ///
+    /// This function uses the fact that the protobuf [`proto::Signed`]
+    /// type has a particular layout when decoding the message to
+    /// efficiently get the bytes that were signed by the signer that
+    /// generated the message. It does the following:
+    /// 1. Decode the first field using the given bytes. This field is the
+    ///    signature field.
+    /// 2. Takes a reference to the bytes after the protobuf signature
+    ///    field. These were supposed to be used generate the digest that
+    ///    was signed over.
+    /// 3. Finish decoding the given bytes into the signed message.
+    /// 4. Transform the protobuf type into the local type.
+    /// 5. Use the local type along with the bytes from (2) to create the
+    ///    signed digest.
+    ///
+    /// One of the assumptions is that repeated serialization of prost
+    /// types generate the same bytes. It also assumes that fields are
+    /// serialized in order by their tag. This is not true for protobufs
+    /// generally, but it is for the prost protobuf implementation. The
+    /// implementation was checked by inspecting the output of `cargo
+    /// expand`.
+    /// <https://protobuf.dev/programming-guides/serialization-not-canonical/>
+    /// <https://protobuf.dev/programming-guides/encoding/#order>
+    /// <https://github.com/tokio-rs/prost/blob/v0.12.6/prost/src/message.rs#L108-L134>
+    pub fn decode_with_digest(data: &[u8]) -> Result<(Self, [u8; 32]), Error> {
```

Comment:
IIUC another assumption here is that when we encode (in `to_digest` below):
```
        let ans = proto::Signed {
            signature: None,
            signer_message: Some(self.clone().into()),
        };
```
we are zero-length encoding `signature` (and `Signed` itself)? I guess this is a proto spec, but it seems we are messing with proto internals to do something we probably shouldn't.
At this point, wouldn't it be better to have `Signed` with `signer_message: bytes` directly? We would need to do a double encoding/decoding ourselves, but we can avoid some assumptions about the layout (and doing prost work).

Reply:
That's a good point. Let me try that.

Reply:
Wait, I'm having cold feet about both approaches.

Reply:
> IIUC another assumption here is that when we encode (in `to_digest` below):
> 
> ```
>         let ans = proto::Signed {
>             signature: None,
>             signer_message: Some(self.clone().into()),
>         };
> ```
> 
> we are zero-length encoding `signature` (and `Signed` itself)? I guess this is a proto spec, but it seems we are messing with proto internals to do something we probably shouldn't.

Yeah, I believe that this assumption is part of the protobuf spec and so should be constant across all implementations.

> At this point, wouldn't it be better to have `Signed` with `signer_message: bytes` directly? We would need to do a double encoding/decoding ourselves, but we can avoid some assumptions about the layout (and doing prost work).

I really want to keep the protobuf definition self describing, but it appears that we pay a steep price for doing so. So I'll try changing it.